src/include/OpenImageIO/refcnt.h

```diff
@@ -39,18 +39,105 @@
 #ifndef OPENIMAGEIO_REFCNT_H
 #define OPENIMAGEIO_REFCNT_H
 
-#include "thread.h"
-#include "oiioversion.h"
+#include <OpenImageIO/thread.h>
+
+#if OIIO_CPLUSPLUS_VERSION >= 11
+# include <memory>
+#else
+# include <boost/shared_ptr.hpp>
+#endif
 
-// Use Boost for shared pointers
-#include <boost/shared_ptr.hpp>
-#include <boost/intrusive_ptr.hpp>
 
 
 OIIO_NAMESPACE_BEGIN
 
+#if OIIO_CPLUSPLUS_VERSION < 11
 using boost::shared_ptr;
-using boost::intrusive_ptr;
+#else
+using std::shared_ptr;
+#endif
+
+
+
+/// A simple intrusive pointer, modeled after std::shared_ptr and
+/// boost::intrusive_ptr.
+template<class T>
+class intrusive_ptr
+{
+public:
+    typedef T element_type;
+
+    /// Default ctr
+    intrusive_ptr () OIIO_NOEXCEPT : m_ptr(NULL) { }
+
+    /// Construct from a raw pointer (presumed to be just now allocated,
+    /// and now owned by us).
+    intrusive_ptr (T *ptr) : m_ptr(ptr) {
+        if (m_ptr) intrusive_ptr_add_ref(m_ptr);
+    }
+
+    /// Construct from another intrusive_ptr.
+    intrusive_ptr (const intrusive_ptr &r) : m_ptr(r.get()) {
+        if (m_ptr) intrusive_ptr_add_ref (m_ptr);
+    }
+
+#if OIIO_CPLUSPLUS_VERSION >= 11
+    /// Move construct from another intrusive_ptr.
+    intrusive_ptr (intrusive_ptr &&r) OIIO_NOEXCEPT : m_ptr(r.get()) {
+        r.m_ptr = NULL;
+    }
+#endif
+
+    /// Destructor
+    ~intrusive_ptr () { if (m_ptr) intrusive_ptr_release (m_ptr); }
+
+    /// Assign from intrusive_ptr
+    intrusive_ptr & operator= (const intrusive_ptr &r) {
+        intrusive_ptr(r).swap (*this);
+        return *this;
+    }
+
+#if OIIO_CPLUSPLUS_VERSION >= 11
+    /// Move assignment from intrusive_ptr
+    intrusive_ptr & operator= (intrusive_ptr&& r) OIIO_NOEXCEPT {
+        intrusive_ptr (static_cast<intrusive_ptr&&>(r)).swap(*this);
```

Comment:
To be honest, being shaky on move semantics myself, I modeled that after what boost did for their intrusive_ptr's. I don't know if it's strictly necessary, but I wasn't sure how to determine whether it was superfluous or not.